[docs] Guide - Connecting Services Across Datacenters (#6052)
hashicorp/consul: website/source/docs/guides/connect-gateways.md

---
layout: "docs"
page_title: "Connecting Services Across Datacenters"
sidebar_current: "docs-guides-connect-gateways"
description: |-
  Connect services and secure inter-service communication across datacenters
  using Consul Connect and mesh gateways.
---
## Introduction

Consul Connect is Consul’s service mesh offering, which allows users to observe
and secure service-to-service communication. Because Connect implements mutual
TLS between services, it also provides users with a way to help services in
different datacenters communicate with each other through mesh gateways. Using
gateways for inter-datacenter communication can prevent each Connect proxy from
needing an accessible address, and frees operators from worrying about IP
overlap between datacenters.

In this guide, you will configure Consul Connect in multiple, joined Consul

> datacenters and use gateways to facilitate inter-service traffic between them.

Reviewer: Everywhere in this document we should probably refer to "gateways" as "mesh gateways". Right now maybe it doesn't matter but if/when we also support API gateways then we will need to differentiate them.

Author: adding a note in the introduction about this. Adding mesh everywhere makes some sentences confusing because it makes nouns really long

Consul clients. Both the gateway and sidecar proxies will need to get
configuration and updates from a local Consul client.

## Enable Connect in Both Datacenters

Once you have your datacenters set up and ACL replication configured, it’s time
to enable Connect in each of them sequentially. Connect’s certificate authority
(which is distinct from the Consul certificate authority that you manage using
the CLI) will automatically bootstrap as soon as a server with Connect enabled
becomes the server cluster’s leader. You can also use [Vault as a Connect
CA](https://www.consul.io/docs/connect/ca/vault.html).

!> If you are using this guide as a production playbook, we strongly recommend
that you enable Connect in each of your datacenters by following the [Connect in
Production
guide](https://learn.hashicorp.com/consul/developer-segmentation/connect-production),
which includes production security recommendations.

### Enable Connect in the primary datacenter

Enable Connect in the primary data center and bootstrap the Connect CA by adding
the following snippet to the server configuration for each of your servers.

```
connect {
  enabled = true
}
```

> Load the new configuration by restarting each server one at a time, making sure

Author: Just to clarify: For this guide, I should just say to "restart Consul" and link out to the rolling restart doc?

> Load the new configuration by restarting each server one at a time, making sure

Reviewer: Yeah, I think thats best

to maintain quorum. Stop the first server by running the following [leave
command](https://www.consul.io/docs/commands/leave.html).

```
$ consul leave
```

Once the server shuts down, restart it with the Consul agent command.

```
$ consul agent -server
```

Once the server is healthy and has rejoined the other servers you can restart
the next one.

### Enable Connect in the secondary datacenter

Once Connect is enabled in the primary datacenter, follow the same process to
enable Connect in the secondary datacenter. Add the following configuration to
the configuration for your servers, and restart them one at a time, making sure
to maintain quorum.

```
connect {
  enabled = true
}
```

> The `primary_datacenter` setting that was required in order to enable ACL

Author: Does it help if I add this note to the prerequisites section? 

-> Following this ACL Replication guide's recommendation for the replication
token policy will work for this guide. The most conservative replication token
policy that will allow intention replication would deny access to service prefixes and
explicitly allow read access to intentions.

> The `primary_datacenter` setting that was required in order to enable ACL

Reviewer: Well the replication token needs different permissions depending on what you want to accomplish. 

* ACL policy/role replication requires `acl:read` 
* ACL token replication requires `acl:write`
* CA replication requires `operator:write`
* Intention replication requires `service:*:read`
* config entry replication requires the same as intention replication right now. 

In general replication tokens should grant pretty broad read access (exception for ACL tokens but those are special)

> The `primary_datacenter` setting that was required in order to enable ACL

Author: Ok, cool. I added some more description to the prerequisites section that I think should explain it better :)

replication between datacenters also specifies which datacenter will act as the
[root CA for Connect](LINK TO MULTI-DC CONNECT DOCS) and write intentions.
Intentions are automatically replicated to the secondary datacenter.

## Deploy Gateways

Connect mesh gateways proxy requests from services in one datacenter to services
in another, so you will need to deploy your gateways on nodes that can reach
each other over the network. You’ll also need to [generate a
token](https://learn.hashicorp.com/consul/security-networking/production-acls#apply-individual-tokens-to-the-services)

> for each gateway that gives it read access to the entire catalog. You’ll apply

Author: Do both gateways need separate tokens? If so, do they need separate policies?

you will need to make sure that both Envoy and Consul are installed on the
gateway nodes. You won’t want to run any services on these nodes other than
Consul and Envoy because they need to be more accessible than the other nodes in
your cluster, and will have access to the WAN.

### Deploy the Gateway for your primary datacenter

Register and start the gateway in your primary datacenter with the following
command.

```
consul connect envoy -mesh-gateway -register \
                     -address "<your private address>" \
                     -wan-address "<your externally accessible address>"\
                     -token=<token for the primary dc gateway>
```

Next, create a [centralized configuration](LINK) file for all the sidecar
proxies in your primary datacenter called `proxy-defaults.hcl`. This file will
instruct the sidecar proxies to send all their inter-datacenter through the
gateway. It should contain the following:

```
Kind = "proxy-defaults"
Name =  "global"
MeshGateway = "local"
```

Write the centralized configuration you just created with the following command.

```
consul config write proxy-defaults.hcl
```

> ### Repeat the Process to Deploy the Gateway for your Secondary Datacenter

Reviewer: This should be done prior to flipping the proxy-defaults switch above, or else any existing cross-dc connect communications are going to temporarily fail until the secondary datacenter has a gateway.

Repeat the above process for the secondary datacenter. Register and start the
gateway in your secondary datacenter with the following command.

```
consul connect envoy -mesh-gateway -register \
                     -address "<your private address>" \
                     -wan-address "<your externally accessible address>"\
                     -token=<token for the secondary dc gateway>
```

> Next, create a centralized configuration file for all the sidecar proxies in

Reviewer: Setting the proxy-defaults config entry only needs to be done once (not per-datacenter).

across multiple datacenters. Now you are ready to register the services that
will use Connect.

## Register a Service in Each Datacenter to Use Connect

You can register a service to use a sidecar proxy by including a sidecar proxy
stanza in its registration file. For this guide, you can use netcat to act as a
backend service and register a dummy service called web to represent the client
service. Those names are used in our examples. If you have services that you
would like to connect, feel free to use those instead.

~> **Caution:** Connect takes its default intention policy from Consul’s default
ACL policy. If you have set your default ACL policy to deny (as is recommended
for secure operation)  and are adding Connect to already registered services,
those services will lose connection to each other until you set an intention
between them to allow communication.

### Register a back end service in one datacenter

In one datacenter register a backend service and add an Envoy sidecar proxy
registration. To do this you will either create a new registration file or edit
an existing one to include a sidecar proxy stanza. If you are using netcat as

> your backend service, you will create a new file called `netcat.json` that would

Author: It looks like most of our docs list json examples so I think its best to use all json instead of all hcl for this guide. Switching everything over 😄

contain the below snippet. Since you have ACLs enabled, you will have to create
a token for the service.

```

> {

Reviewer: I don't see where we tell them the netcat/socat command to run that listens on port 8181? Are we assuming they already know what to do from the first connect guide?

Reload the client with the new or modified registration.

```
consul reload
```

Then start Envoy specifying which service it will proxy.

```
consul connect envoy -sidecar-for netcat
```

If you are using netcat as your example, start it now by running the following
command. netcat will run in your terminal window and echo back anything you type
on the command line.

```
$ nc 127.0.0.1 8181
hello
hello
echo
echo
```

Stop the running service by typing `ctrl + c`.

### Register a front end service in the other datacenter

Now in your other datacenter, you will register a service that calls your
backend service, with a sidecar proxy Your registration will need to list the
backend service as your upstream. Like the backend service, you can use our
example service, which will be called web, or append the connect stanza to an
existing registration with some customization.

To use web as your front end service, create a registration file called
`web.json` that contains the following snippet.

```
{
  "service": {
    "name": "web",
    "port": 8080,
    "token": <token here>,
    "connect": {
      "sidecar_service": {
        "proxy": {
          "upstreams": [{

>              "destination_name": "netcat",

Reviewer: The upstream config also needs a datacenter argument or else they will look in the local datacenter for the service:

```
upstreams = [{
  destination_name = "netcat"
  datacenter = "primary"
  local_bind_port = 8181
}]
```

upstream. If you are using another service as a back end, replace `socat` with
its name and the `8181` with its port.

Reload the client with the new or modified registration.

```text
$ consul reload
```

Then start Envoy and specify which service it will proxy.

```text
$ consul connect envoy -sidecar-for web
```

## Configure Intentions to Allow Communication Between Services

Now that your services both use Connect, you will need to configure intentions
in order for them to communicate with each other. Add an intention to allow the
front end service to access the back end service. For web and socat the command
would look like this.

```text
$ consul intention create web socat
```

Consul will automatically forward intentions initiated in the in the secondary
datacenter to the primary datacenter, where the servers will write them. The
servers in the primary datacenter will then automatically replicate the written
intentions back to the secondary datacenter.

## Test the connection

Now that you have services using Connect, verify that they can contact each
other. If you have been using the example web and socat services, from the node

> and datacenter where you registered the web service, start the socat service

Reviewer 2: This is almost there, needs one more adjustment. The socat service is already running upstream, so to test web's connection to socat we use netcat:
>If you have been using the example web and socat services, from the node and datacenter where you registered the web service, use netcat to test the connection to the socat echo server.

Author: Good catch, thank you!

---
layout: "docs"
page_title: "Connecting Services Across Datacenters"
sidebar_current: "docs-guides-connect-gateways"
description: |-
  Connect services and secure inter-service communication across datacenters
  using Consul Connect and mesh gateways.
---
## Introduction

Consul Connect is Consul’s service mesh offering, which allows users to observe
and secure service-to-service communication. Because Connect implements mutual
TLS between services, it also enabled us to build mesh gateways, which provide
users with a way to help services in different datacenters communicate with each
other. Mesh gateways take advantage of Server Name Indication (SNI), which is an
extension to TLS that allows them to see the destination of inter-datacenter

> traffic without de-encrypting the message payload.

Reviewer: ```suggestion
traffic without decrypting the message payload.
```

> Using gateways for inter-datacenter communication can prevent each Connect proxy

Reviewer: ```suggestion
Using mesh gateways for inter-datacenter communication can prevent each Connect proxy
```

Or you could move up the bit below that clarifies that within this doc "gateway" means mesh gateway

1. Test that your services can communicate with each other

For the remainder of this guide we will refer to mesh gateways as "gateways".
Anywhere in this guide where you see the word gateway, assume it is specifically
a mesh gateway (as opposed to an API or other type of gateway).

## Prerequisites

To complete this guide you will need two wide area network (WAN) joined Consul
datacenters with access control list (ACL) replication enabled. If you are
starting from scratch, follow these guides to set up your datacenters, or use
them to check that you have the proper configuration:

- [Deployment Guide](/consul/datacenter-deploy/deployment-guide)
- [Securing Consul with ACLs](/consul/security-networking/production-acls)
- [Basic Federation with WAN Gossip](/consul/security-networking/datacenters)

You will also need to enable ACL replication, which you can do by following the [ACL Replication for Multiple Datacenters](/consul/day-2-operations/acl-replication) guide with the following modification.

When creating the [replication token for ACL
management](/consul/day-2-operations/acl-replication#create-the-replication-token-for-acl-management),
it will need the following policy:

> ```json

Reviewer: The HCL form of the policy would look much nicer.

```hcl
acl = "write"
operator = "write"
service_prefix "" {
   policy = "read"
}
```

Also of note is that its a service_prefix rule not a service rule.

Author: I think @freddygv's suggestion was to stay consistent with the language for all our snippets, which is why I decided to convert it. Probably worth some thought about whether or not we care about being consistent. I think the ACL docs are one of the only places in the documentation where we use HCL, perhaps because the JSON gets extra ugly there.

{
  "acl": "write",
  "operator": "write",

>   "service": {

Author: ```suggestion
  "service_prefix": {
```

    "" : {
      "policy": "read"
    }
  }
}
```

The replication token needs different permissions depending on what you want to

> accomplish. The above policy enables ACL policy, role, and token replication

Reviewer: ```suggestion
accomplish. The above policy allows for ACL policy, role, and token replication
```

  -policy-name mesh-gateway
```

```text
$ consul acl token create \
  -description "mesh-gateway secondary datacenter token" \
  -policy-name mesh-gateway
```

You’ll apply those tokens when you deploy the gateways.

### Deploy the Gateway for your primary datacenter

Register and start the gateway in your primary datacenter with the following
command.

```text
$ consul connect envoy -mesh-gateway -register \
                     -service-name "gateway-primary"
                     -address "<your private address>" \
                     -wan-address "<your externally accessible address>"\
                     -token=<token for the primary dc gateway>
```

### Deploy the Gateway for your Secondary Datacenter

Register and start the gateway in your secondary datacenter with the following
command.

```text
$ consul connect envoy -mesh-gateway -register \
                     -service-name "gateway-secondary"
                     -address "<your private address>" \
                     -wan-address "<your externally accessible address>"\
                     -token=<token for the secondary dc gateway>
```

> ### Configure Sidecar Proxies to use Gateways

Reviewer: One thing to note is that for this to work you have to set `enable_central_service_config = true` in all the configuration of agents that will be managing sidecar proxies (although it would also be useful for all agents even ones only managing gateways).

```text
$ consul connect envoy -sidecar-for socat
```

If you are using socat as your example, start it now on the port you specified
in your registration by running the following command.

```text
$ socat -v tcp-l:8181,fork exec:"/bin/cat"
```

Check that the socat service is running by accessing it using netcat on the same
node. It will echo back anything you type.

```text
$ nc 127.0.0.1 8181
hello
hello
echo
echo
```

Stop the running netcat service by typing `ctrl + c`.

### Register a front end service in the other datacenter

Now in your other datacenter, you will register a service (with a sidecar proxy)
that calls your backend service. Your registration will need to list the backend
service as your upstream. Like the backend service, you can use an example
service, which will be called web, or append the connect stanza to an existing
registration with some customization.

To use web as your front end service, create a registration file called
`web.json` that contains the following snippet.

```json
{
  "service": {
    "name": "web",
    "port": 8080,
    "token": "<token here>",
    "connect": {
      "sidecar_service": {

>         "proxy": {

Reviewer: Per my other comment above, if you don't enable central configuration you can set up gateways per proxy/upstream like:

Per Proxy
```json
{ 
   "proxy": {
      "mesh_gateway": {
         "mode": "local"
      }
      ...
   }
}
```

Per Upstream
```json
{ 
   "proxy": {
      "upstreams": [
         {
            "mesh_gateway": {
               "mode": "local"
            },
            ...
         },
         ...
      ]
      ...
   }
}
